File: lib/languages/kotlin/generics.kt
// Generic function
fun <T> identity(value: T): T {
   return value
}

// Using a generic function
val stringResult: String = identity("hello")
val intResult: Int = identity(42)

// Generic class
class Box<T>(var value: T) {
   fun getValue(): T {
       return value
   }
   
   fun setValue(newValue: T) {
       value = newValue
   }
}

// Using a generic class
val stringBox = Box("hello")
val intBox = Box(42)

// Multiple type parameters
class Pair<A, B>(val first: A, val second: B)

// Type constraints
fun <T : Comparable<T>> findMax(list: List<T>): T? {
   if (list.isEmpty()) return null
   var max = list[0]
   for (item in list.drop(1)) {
       if (item > max) max = item
   }
   return max
}

// Multiple constraints with where clause
fun <T> copyWhenGreater(list: List<T>, threshold: T): List<String>
       where T : CharSequence,
             T : Comparable<T> {
   return list.filter { it > threshold }.map { it.toString() }
}

// Type projection (covariance)
fun copyStrings(from: Array<out String>, to: Array<in String>) {
   // 'out' makes 'from' covariant (producer)
   // 'in' makes 'to' contravariant (consumer)
}

// Star projection
fun printList(list: List<*>) {
   list.forEach { println(it) }
}

// Reified type parameters (with inline)
inline fun <reified T> isA(value: Any): Boolean {
   return value is T
}

// Generic extension function
fun <T> T.toBox(): Box<T> {
   return Box(this)
}

// Type erasure workaround with reified
inline fun <reified T> List<*>.filterIsInstance(): List<T> {
   val result = mutableListOf<T>()
   for (element in this) {
       if (element is T) {
           result.add(element)
       }
   }
   return result
}

File: lib/languages/rust/generics.rs
// Generic function
fn identity<T>(x: T) -> T {
   x
}

// Using a generic function
let int_value = identity(42);
let string_value = identity("hello");

// Generic struct
struct Box<T> {
   value: T,
}

// Using a generic struct
let int_box = Box { value: 42 };
let string_box = Box { value: String::from("hello") };

// Generic enum
enum Option<T> {
   Some(T),
   None,
}

// Generic function with multiple type parameters
fn pair<T, U>(first: T, second: U) -> (T, U) {
   (first, second)
}

// Generic function with trait bounds
fn print<T: std::fmt::Display>(value: T) {
   println!("{}", value);
}

// Multiple trait bounds
fn sort_and_print<T: Ord + std::fmt::Debug>(mut collection: Vec<T>) -> Vec<T> {
   collection.sort();
   println!("{:?}", collection);
   collection
}

// Where clauses for complex bounds
fn process<T, U>(t: T, u: U) -> bool
where
   T: std::fmt::Display + Clone,
   U: std::fmt::Debug + PartialEq,
{
   println!("{}", t);
   println!("{:?}", u);
   true
}

// Generic implementations
impl<T> Box<T> {
   fn new(value: T) -> Box<T> {
       Box { value }
   }
   
   fn get(&self) -> &T {
       &self.value
   }
}

